### ISSUE TITLE
[BUG] Corrupted data from HyperplaneGenerator when batch_size > 1

### ISSUE THREAD

**Issue by zhanghaooo** [scikit-multiflow/scikit-multiflow#135]
Hi, I save the data that generated from a stream generator to a .csv file, and create a FileStream by reading the file, but what strange is that the evaluation result about the FileStream looks abnormal. Here is my code:

data_generator.py
```python
stream = HyperplaneGenerator(random_state=666, n_features=2, n_drift_features=0, mag_change=0.0, noise_percentage=0.0, sigma_percentage=0.0)
stream.prepare_for_use()
total_sample = 100000
X, y = stream.next_sample(total_sample)
df_X = pd.DataFrame(np.array(X).astype('float'))
df_y = pd.DataFrame(np.array(y).astype('int').T)
pd.concat([df_X, df_y], axis=1).to_csv("data.csv", index=False, header=False)
```
```python
from skmultiflow.data.file_stream import FileStream
from skmultiflow.trees import HoeffdingTree

stream = FileStream("data.csv")
'''
stream = HyperplaneGenerator(random_state=666, n_features=2, n_drift_features=0, mag_change=0.0, noise_percentage=0.0, sigma_percentage=0.0)
# But if I use HyperplaneGenerator() directly at here, the evaluation result looks fine.
'''
stream.prepare_for_use()
evaluator = EvaluatePrequential(show_plot=True,
                                max_samples=1000000,
                                metrics=['precision', 'recall', 'f1'],
                                data_points_for_classification=False)
evaluator.evaluate(stream=stream, model=HoeffdingTree())
```

Could anyone help me? Thanks a lot~~~

**Comment by jacobmontiel** [comment]
Hello @zhanghaooo, could you elaborate on "the evaluation result about the FileStream looks abnormal"?

**Comment by zhanghaooo** [comment]
Ok, when I use the HyperplaneGenerator() directly, and evaluator the prequential performance of a HoeffdingTree on the stream, the result looks fine:
![scikit-multiflow1](https://user-images.githubusercontent.com/35730721/64219609-b9725580-cef8-11e9-8a65-7d2cda431c34.png)
But if I save the stream to a file, and then create a stream by FileStream(), the evaluation result looks abnormal:
![scikit-multiflow](https://user-images.githubusercontent.com/35730721/64219670-f76f7980-cef8-11e9-8cf4-f11735000d43.png)
I think that the result should be the same because of the same data stream, isn't it?

**Comment by jacobmontiel** [comment]
Hello @zhanghaooo 

I have not been able to replicate the error. Are you sure you are saving/loading the file in the same path? Could you try removing and recreating the file in case it is corrupted?

Comment, your example uses a different number of sample. The file has **100K** samples while the part with the generator uses **1M** samples.

Here are the results I got on 100K:

**From generator**
![from_generator](https://user-images.githubusercontent.com/17923265/64226475-a6c94200-cf33-11e9-8aab-2dc6a0f0e76a.png)

**From file**
![from_file](https://user-images.githubusercontent.com/17923265/64226482-ab8df600-cf33-11e9-8c9f-9f777e7a749c.png)

**Comment by zhanghaooo** [comment]
Hello @jacobmontiel 

This time I put all the code in a file:
```python
from skmultiflow.data import FileStream
from skmultiflow.trees import HoeffdingTree
from skmultiflow.evaluation import EvaluatePrequential
from skmultiflow.data import HyperplaneGenerator
import numpy as np
import pandas as pd

stream = HyperplaneGenerator(random_state=666, n_features=2, n_drift_features=0,
                             mag_change=0.0, noise_percentage=0.0, sigma_percentage=0.0)
stream.prepare_for_use()
total_sample = 100000
X, y = stream.next_sample(total_sample)
df_X = pd.DataFrame(np.array(X).astype('float'))
df_y = pd.DataFrame(np.array(y).astype('int').T)
pd.concat([df_X, df_y], axis=1).to_csv("data.csv", index=False, header=False)

stream = FileStream("data.csv")
'''
# But if I use HyperplaneGenerator() directly at here, the evaluation result looks fine.
stream = HyperplaneGenerator(random_state=666, n_features=2, n_drift_features=0, 
                             mag_change=0.0, noise_percentage=0.0, sigma_percentage=0.0)
'''
stream.prepare_for_use()
evaluator = EvaluatePrequential(show_plot=True,
                                max_samples=total_sample,
                                metrics=['precision', 'recall', 'f1'],
                                data_points_for_classification=False)
evaluator.evaluate(stream=stream, model=HoeffdingTree())
```
Unfortunately, the results still looks like this:
![scikit-multiflow](https://user-images.githubusercontent.com/35730721/64227179-1419aa80-cf15-11e9-933d-34ce31bc9b6b.png)
but if I cancel the comment to use the HyperplaneGenerator() directly, the results looks fine:

![scikit-multiflow](https://user-images.githubusercontent.com/35730721/64227300-84c0c700-cf15-11e9-92af-8a89c18c3659.png)

**Comment by zhanghaooo** [comment]
Hello @jacobmontiel 

Oh, I find a solution! If I generate the sample one by one instead of 10K samples at one time. 
```Python
from skmultiflow.data import FileStream
from skmultiflow.trees import HoeffdingTree
from skmultiflow.evaluation import EvaluatePrequential
from skmultiflow.data import HyperplaneGenerator
import numpy as np
import pandas as pd

stream = HyperplaneGenerator(random_state=666, n_features=2, n_drift_features=0,
                             mag_change=0.0, noise_percentage=0.0, sigma_percentage=0.0)
stream.prepare_for_use()
n_features = 2
total_sample = 100000
X, y = stream.next_sample(total_sample)
X = np.zeros(shape=(total_sample, n_features)).astype('float')
y = np.zeros(shape=(total_sample, 1)).astype('int')
for i in range(total_sample):
    X[i, :], y[i, 0] = stream.next_sample()

df_X = pd.DataFrame(X)
df_y = pd.DataFrame(y)
pd.concat([df_X, df_y], axis=1).to_csv("data.csv", index=False)

stream = FileStream("data.csv")
stream.prepare_for_use()
evaluator = EvaluatePrequential(show_plot=True,
                                max_samples=total_sample,
                                metrics=['precision', 'recall', 'f1'],
                                data_points_for_classification=False)
evaluator.evaluate(stream=stream, model=HoeffdingTree())
```
the result will be right:
![scikit-multiflow](https://user-images.githubusercontent.com/35730721/64227577-9060bd80-cf16-11e9-8d04-280a893883d4.png)
But it confused me that getting sample one by one from a stream is different from get a lot of samples from a stream at once?

**Comment by zhanghaooo** [comment]
Hello@jacobmontiel 

By the way, I noted that the titles of figures show 'Unknown metric' when I use some kinds of scores such as 'precision','recall','f1' and so on. Just now, I found that the problem have been fixed https://github.com/scikit-multiflow/scikit-multiflow/blob/2dadabe68d99e32ce40e841f6790aacdbaa758d1/src/skmultiflow/visualization/evaluation_visualizer.py#L276
But it seems like I can't get the update by pip, even when I uninstall it and install again. Do I have to somplie the src if I want to get the update?

**Comment by jacobmontiel** [comment]
I ran your last example where 100K samples are requested and stored into the file and got the right (expected) result. I am not really sure why it is different for you. As you mention, there is no difference if you generate one sample at a time or `n` samples in a single call.

The fix you mention is available in the development branch of the project which is not reachable through pip. We will be publishing the last release either this or next week that will include this and other changes.

**Comment by zhanghaooo** [comment]
Hello@jacobmontiel
I found that the data generated from a HyperplaneGenerator() stream one by one is indeed different from generating a lots samples from the same stream one time.

Here is a jupyter notebook record: 
```python
from skmultiflow.data import HyperplaneGenerator
import numpy as np
stream = HyperplaneGenerator(random_state=666, n_features=2)
stream.prepare_for_use()
total_sample = 10
```

```python
X, y = stream.next_sample(total_sample)
X, y
```
    (array([[0.67651434, 0.72785806],
            [0.0127032 , 0.4135877 ],
            [0.09992856, 0.50806631],
            [0.74415417, 0.192892  ],
            [0.29322811, 0.77447945],
            [0.11285765, 0.11095367],
            [0.0232363 , 0.72732115],
            [0.19750316, 0.90917959],
            [0.53280254, 0.25913185],
            [0.32569065, 0.88889931]]), array([1, 0, 0, 0, 1, 0, 0, 0, 0, 0]))

```python
stream.restart()
X = np.zeros(shape=(total_sample,2))
y = np.zeros(shape=(total_sample,)).astype('int')
for i in range(total_sample):
    X[i,:], y[i] = stream.next_sample()
X, y
```
    (array([[0.67651434, 0.72785806],
            [0.04881279, 0.09992856],
            [0.192892  , 0.70084475],
            [0.11285765, 0.11095367],
            [0.34003494, 0.19750316],
            [0.25913185, 0.58381262],
            [0.81887369, 0.54734542],
            [0.07516654, 0.77519298],
            [0.1536739 , 0.82846513],
            [0.90238039, 0.85178834]]), array([1, 0, 0, 0, 0, 0, 1, 0, 1, 1]))

and if I get 100 samples one time, the y vector will be:
```python
stream.restart()
X, y = stream.next_sample(100)
y
```
```python
array([1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0])
```
There are many continuoues 0 and 1. I think this kind of data may not be normal, isn't it?

**Comment by jacobmontiel** [comment]
I realize now what is happening. There is a bug in the latest version (see #129). The fix is only available in the development version to be released. Nonetheless while checking this I found that the fix was incomplete since it was not covering the cases when drift is added. I have uploaded the complement of the fix in c1704a9dc10e2288317464979b8ea556164bbc2f.

A temporary workaround is to use a 'for' loop as you re currently doing. Next release will include the fix to this. Sorry for the inconvenience.
